### PR TITLE
Ros2 reenable tests

### PR DESCRIPTION
**Pull request by mikaelarguedas**
Opening for visibility.

Apparently class_loader tests have been disabled in ROS2 for a looong time (https://github.com/ament/ament_cmake/pull/68). This aims to reenable and fix them

### CODE REVIEW

File: test/fviz_case_study/CMakeLists.txt [hunk 1]
```diff
@@ -17,15 +17,26 @@ add_library(${PROJECT_NAME}_Test_FvizDefaultPlugin
...
   PUBLIC "../include" ${console_bridge_INCLUDE_DIRS} ${Poco_INCLUDE_DIRS})
 target_link_libraries(${PROJECT_NAME}_Test_FvizDefaultPlugin
-  ${PROJECT_NAME} ${PROJECT_NAME}_Test_Fviz)
+  ${PROJECT_NAME} ${PROJECT_NAME}_Test_Fviz ${console_bridge_LIBRARIES} ${Poco_LIBRARIES})
 class_loader_hide_library_symbols(${PROJECT_NAME}_Test_FvizDefaultPlugin)
 
-ament_add_gtest(${PROJECT_NAME}_fviz_test fviz_test.cpp)
+set(append_library_dirs "${CMAKE_CURRENT_BINARY_DIR}")
+if(WIN32)
+  set(append_library_dirs "${append_library_dirs}/../../$<CONFIG>;${append_library_dirs}/$<CONFIG>")
```
Comment (mikaelarguedas) [inline test/fviz_case_study/CMakeLists.txt]: @wjwwood @dirk-thomas this is pretty ugly. Am I approaching this the wrong way?

File: src/class_loader.cpp [hunk 2]
```diff
@@ -56,12 +56,17 @@ std::string systemLibraryPrefix()
...
+// to automatically remove the trailing `d` from the shared library suffix
+//   return Poco::SharedLibrary::suffix();
+#ifdef __linux__
+  return ".so";
+#elif __APPLE__
+  return ".dylib";
+#elif _WIN32
   return ".dll";
+#else
+  return Poco::SharedLibrary::suffix();
```
Comment (dirk-thomas) [inline src/class_loader.cpp]: Wouldn't it be nice if the debug shared libraries could be loaded for debugging?

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -7,43 +7,62 @@ target_include_directories(${PROJECT_NAME}_TestPlugins1
   PUBLIC "../include" ${console_bridge_INCLUDE_DIRS} ${Poco_INCLUDE_DIRS})
 target_link_libraries(${PROJECT_NAME}_TestPlugins1 ${PROJECT_NAME})
 class_loader_hide_library_symbols(${PROJECT_NAME}_TestPlugins1)
-
+if(WIN32)
+target_compile_definitions(${PROJECT_NAME}_TestPlugins1 PRIVATE "CLASS_LOADER_BUILDING_DLL")
```
Comment (dirk-thomas) [inline test/CMakeLists.txt]: My expectation would be that this should only be defined when `class_loader` is being built. Why is this necessary here? And if it is I would suggest to use a separate name to indicate that the "using" part of class loader is being built here.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -7,43 +7,62 @@ target_include_directories(${PROJECT_NAME}_TestPlugins1
...
+    ${PROJECT_NAME}_TestPlugins2
+    ${Poco_LIBRARIES}
+  )
   if(NOT WIN32)
     target_link_libraries(${PROJECT_NAME}_utest pthread)
   endif()
-  add_dependencies(${PROJECT_NAME}_utest ${PROJECT_NAME}_TestPlugins1 ${PROJECT_NAME}_TestPlugins2)
+  add_dependencies(${PROJECT_NAME}_utest
+    ${PROJECT_NAME}_TestPlugins1
+    ${PROJECT_NAME}_TestPlugins2)
```
Comment (dirk-thomas) [inline test/CMakeLists.txt:39]: Since the target already links against these libraries the dependency declaration is redundant.

Same below.

File: test/fviz_case_study/CMakeLists.txt (hunk 1 above)
Comment (dirk-thomas) [reply test/fviz_case_study/CMakeLists.txt]: I get why you want to use `${append_library_dirs}/$<CONFIG>` but where is `${append_library_dirs}/../$<CONFIG>` supposed to be pointing to?

Maybe you can use a generator expression to point to the target directory of one of the plugin libraries instead  (like `$<TARGET_FILE_DIR:${PROJECT_NAME}_TestPlugins1>`)?

File: test/CMakeLists.txt (hunk 3 above)
Comment (mikaelarguedas) [reply test/CMakeLists.txt]: Without this, the tests linking against these testing libraries fail to link.

>  And if it is I would suggest to use a separate name to indicate that the "using" part of class loader is being built here.

I'm not sure I get what you mean here, do you suggest creating another set of visibility macros for the libraries built and used for testing?

File: src/class_loader.cpp (hunk 2 above)
Comment (mikaelarguedas) [reply src/class_loader.cpp]: Yes that would be great, I'll track this on https://github.com/ros/class_loader/issues/31 given that this commit already got merged in #60 
(commenting now because the rebase + force push may overwrite this comment)

File: test/CMakeLists.txt (hunk 4 above)
Comment (mikaelarguedas) [reply test/CMakeLists.txt:39]: indeed the target should not link these libraries, fixed in 08d4345335c1b4b64e4761a490e3136ba87d36c4

**Comment by mikaelarguedas** [comment]
Ok I think this is ready for a round of reviews

This PR reenables the tests for ROS2, convert the tests using catkin and boost to use ament and STL and modifies test amd CMake code to pass on all platforms

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3458/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/699/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2787/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/3505/)

Note: I removed all the checks for c++11 support, my assumption here is that I'll branch out class_loader for Melodic and will not need that check in Melodic either because all compilers will support c++11 by default. I will reconciliate the melodic-devel and the ros2 branches at that point.

**Review by wjwwood — APPROVED** [review]
lgtm, thanks for cleaning up those tests for me

**Comment by mikaelarguedas** [comment]
ok I rerun a full set of CI and everything's green, merging
